### PR TITLE
Serialization improvements:

### PR DESCRIPTION
**Pull request by toly-kournik**
1. Serializable
	- Removed usage of tmp files for objects serialization;
	- introduced template helper functions for integer types ser/deser;
	- final serialize/deserialize in Serializable for consistent behavior;
	- symmetric work with serializeDataMembers/deserializeDataMembers;
	- consistent treatment of inheritance;
	- logging;
	- rename SharedPtrToClass => SerializablePtr;
	- rename namspace concordSerializable => serialize.
2. TODO
	- FileStorage to work with ifstream;
	- remove Serializable::deserialize(const UniquePtrToChar &inBuf, int64_t inBufSize)
	- remove MemoryBasedStream and MemoryBasedBuf.
	- more efficient class registration mechanism.

### CODE REVIEW

File: bftengine/tests/simpleStorage/FileStorage.cpp [hunk 1]
```diff
@@ -26,9 +26,10 @@ namespace bftEngine {
 FileStorage::FileStorage(Logger &logger, const string &fileName) :
     logger_(logger), fileName_(fileName) {
 
-  dataStream_ = fopen(fileName.c_str(), "wxb+");
```
Comment (andrewjstone) [inline bftengine/tests/simpleStorage/FileStorage.cpp]: why remove the binary flag?

File: bftengine/tests/simpleStorage/FileStorage.cpp [hunk 2]
```diff
@@ -87,7 +88,19 @@ void FileStorage::writeFileMetadata() {
 
 void FileStorage::read(void *dataPtr, size_t offset, size_t itemSize, size_t count, const char *errorMsg) {
   fseek(dataStream_, offset, SEEK_SET);
-  if (fread(dataPtr, itemSize, count, dataStream_) != count) {
+  size_t read_ = fread(dataPtr, itemSize, count, dataStream_);
+  int err = ferror(dataStream_);
+  if (err)
+  {
```
Comment (andrewjstone) [inline bftengine/tests/simpleStorage/FileStorage.cpp]: nit: brace on prior line per code standard

File: threshsign/src/bls/relic/BlsPublicParameters.cpp [hunk 3]
```diff
@@ -45,6 +46,18 @@ BlsPublicParameters::BlsPublicParameters(const BlsPublicParameters &params)
...
 }
+BlsPublicParameters& BlsPublicParameters::operator=(const BlsPublicParameters& params){
+  securityLevel_ = params.securityLevel_;
+  schemeName_    = params.schemeName_;
+  library_       = params.library_;
+  curveType_     = params.curveType_;
+  BLS::Relic::Library::Get();
+  g1_get_gen(generator1_);
+  g2_get_gen(generator2_);
+//  g1_copy(generator1_, params.generator1_);
```
Comment (andrewjstone) [inline threshsign/src/bls/relic/BlsPublicParameters.cpp]: Wy are these lines commented out? Let's delete them if they aren't needed please.

File: bftengine/tests/simpleStorage/FileStorage.cpp (hunk 1 above)
Comment (toly-kournik) [reply bftengine/tests/simpleStorage/FileStorage.cpp]: The mode string can also include the letter ``b'' after either the ``+'' or the first letter.  This is strictly for compatibility with ISO/IEC 9899:1990 (``ISO C90'') and has effect only for fmemopen() ; other-
     wise ``b'' is ignored.

File: bftengine/src/bftengine/PersistentStorageImp.cpp [hunk 4]
```diff
@@ -16,7 +16,7 @@
 #include <sstream>
 
 using namespace std;
-using namespace concordSerializable;
+using namespace serialize;
```
[the remark this answers is not in the document]
Comment (toly-kournik) [reply bftengine/src/bftengine/PersistentStorageImp.cpp]: I agree

File: threshsign/src/bls/relic/BlsPublicParameters.cpp (hunk 3 above)
Comment (toly-kournik) [reply threshsign/src/bls/relic/BlsPublicParameters.cpp]: I'll do it

File: bftengine/tests/simpleStorage/FileStorage.cpp (hunk 2 above)
Comment (toly-kournik) [reply bftengine/tests/simpleStorage/FileStorage.cpp]: done

File: bftengine/src/bftengine/PersistentStorageImp.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (toly-kournik) [reply bftengine/src/bftengine/PersistentStorageImp.cpp]: done

**Comment by andrewjstone** [comment]
LGTM! Thanks for the fixes Toly. After running the tests as Yulia describes we are good to merge.

**Comment by toly-kournik** [comment]
I have an additional commit to do this eve

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Pull request merged**